soroban-env-host/src/preflight.rs

```diff
@@ -0,0 +1,314 @@
+mod ledger_storage;
+mod fees;
+mod state_ttl;
+
+use anyhow::{anyhow, bail, Context, Result};
+use ledger_storage::LedgerStorage;
+use crate::auth::RecordedAuthPayload;
+use crate::budget::Budget;
+use crate::events::Events;
+use crate::storage::Storage;
+use crate::xdr::{
+    AccountId, ConfigSettingEntry, ConfigSettingId, DiagnosticEvent, InvokeHostFunctionOp,
+    LedgerFootprint, LedgerKey, OperationBody, ScVal, SorobanAddressCredentials,
+    SorobanAuthorizationEntry, SorobanCredentials, SorobanTransactionData, VecM,
+};
+use crate::{DiagnosticLevel, Host, LedgerInfo};
+use std::collections::HashSet;
+use std::convert::{TryFrom, TryInto};
+use std::iter::FromIterator;
+use std::rc::Rc;
+
+pub struct RestorePreamble {
+    pub(crate) transaction_data: SorobanTransactionData,
+    pub(crate) min_fee: i64,
+}
+
+#[derive(Default)]
+pub struct PreflightResult {
+    pub error: String,
+    pub auth: Vec<SorobanAuthorizationEntry>,
+    pub result: Option<ScVal>,
+    pub transaction_data: Option<SorobanTransactionData>,
+    pub min_fee: i64,
+    pub events: Vec<DiagnosticEvent>,
+    pub cpu_instructions: u64,
+    pub memory_bytes: u64,
+    pub restore_preamble: Option<RestorePreamble>,
+}
+
+pub fn preflight_invoke_hf_op(
+    ledger_storage: LedgerStorage,
+    bucket_list_size: u64,
+    invoke_hf_op: InvokeHostFunctionOp,
+    source_account: AccountId,
+    ledger_info: LedgerInfo,
```

Comment:
I'm not certain what exactly do we achieve with this change (besides changing code ownership, I guess?). If that's just the first step before more refactoring, that's fine, but I think even in this scenario I'm not sure if all the logic here belongs to the host, like `LedgerStorage` wrapper.
Basically, on a high level I see the following components:
- Invoking host fn e2e in recording storage mode. This is similar to what `e2e_invoke` already does and it would be nice to share as much code as possible with it. The interface should probably be more abstract (e.g. use `SnapshotSource` instead of `LedgerStorage`). Budget probably should be set up externally. Restoration preamble also probably doesn't belong here. Basically, the closer we can make this to `e2e_invoke::invoke_host_function`, the better.
- Then we could have a module that computes the resource requirements and fee given the result of e2e invocation
- Restore/extend ops preflight really have nothing to do with the host. I don't have a strong opinion as to where they should belong, the main point is that they belong to a separate module.
- XDR config settings are normally abstracted away from host as well. Again, no matter the final location, these probably should belong to a separate module
- The same goes for transaction XDR manipulation - probably should be a separate module
- The 'glue' layer that puts everything together belongs to soroban-rpc. I can't quickly tell what belongs there, but it seems to me that more bespoke stuff like archived entry set for restoration should be there as well.
Basically, we want to encapsulate the host-related logic, so that it's possible to reasonably test it and compare with our e2e_invoke results. Everything else should be separated and covered separately. Again, we don't have to do all the refactoring immediately, but maybe it's worth to lay some foundations, at least in terms of the interface.

Reply:
In general, I agree with your comments. The initial state of this PR mostly had the intention of starting a conversation.

Some remarks/questions:
* I agree that the restore/extend fee computation is mostly unrelated to the host. However, the code should live in a place:
  * Which makes it easy to update whenever Core's metering is updated
  * Where can be automatically tested against Core's metering to make sure they don't diverge.
  * It can share code with the invoke host function fee computation (this is less important).
  Any suggestions?
  
* Maybe we don't need `LedgerStorage` in its current state, but `SnapshotSource`
  1. doesn't allow querying non-live entries which is needed for:
      * the restore operation preflight
      * tracking ledger entries which require restoring. This could arguably stay in soroban-rpc, though
  2. The `Rc`s make it awkward to use
 * Do you have any suggestions as to how to split the code in separate modules other than host? I don't have any strong opinions and I am not familiar with the existing env modules, so it would make the migration smoother is I was guided with respect to that.

Reply:
> Which makes it easy to update whenever Core's metering is updated

I don't think there is a good solution for now, just because core code lives in C++ and this lives in Rust. Moving these to Rust is probably infeasible - the fee computation is already on the Rust side, but doing an XDR roundtrip just to compute the entry size seems inefficient. That said, the Core metering update is a protocol change, so it would happen very rarely (if ever). So while I agree it would be nice to share even more code, the changes are unlikely enough now to not worry about that too much. The same goes for comparison with Core - while it would be nice to have an automated way to test this, it seems like *a lot* of setup and maintenance work, all for very little benefit. I think a compromise solution would be to implement unit tests with the same setup in Core and in the Rust library and leave the reference in Core stating that Rust test has to be in sync with it.

 To be clear, I'm not opposed to storing this in `rs-soroban-env`, but it would probably be cleaner if we had a separate crate for non-host-related stuff like this. I think something like `soroban-preflight` or `soroban-simulation`  would be appropriate if we also move more preflight-specific utils there; otherwise something like `soroban-protocol-utils` would work as well.

> Maybe we don't need LedgerStorage in its current state, but SnapshotSource

To be clear, what I'm suggesting is to just take `SnapshotSource` trait in e2e invocation function; it seems like we need only `SnapshotSource` functions there (besides footprint restoration preamble, but as I've mentioned, I think it should happen outside of this function). `LedgerStorage` itself may live either in `soroban-preflight` in env, or in `soroban-rpc`. This will help with setting up the comparison tests between the 'canonical' and preflight host fn invocations.

> Do you have any suggestions as to how to split the code in separate modules other than host?

As I've suggested above, I would probably move most of the new code into a separate crate. You can then make a pass to maybe separate this into a bit more granular modules (I had some ideas above), but really for the first iteration we don't need to spend too much time on this - we can refactor this further later. The end goal would be to have a reasonably small public interface and a good unit test set that covers it, as well as some tests for internal modules where appropriate.

Reply:
Just a small clarification: I've mentioned 'separate workspace' above, but in fact I meant just a separate workspace member (and respective separate crate) in rs-soroban-env workspace.

Reply:
@dmkozh PTAL, I moved everything to a another workspace member (`soroban-preflight`).

I think it's a good starting point.

Reply:
@dmkozh if you are OK with the structure, I can move on to remove the anyhow/thiserror dependencies.

Reply:
There is some portion of this that we'll need to move to env, but it's probably ok to do as the next step. I'll need to spend some time refactoring this to reuse the env code more, and maybe move some bits back to the soroban-rpc repo if they seem too specific. But this looks fine for now.

Reply:
OK, great! I will tweak the PR a bit further and let you know once it’s ready to be merged.